src/org/rascalmpl/interpreter/TraversalEvaluator.java

```diff
@@ -345,13 +347,25 @@ private IValue traverseADTOnce(IValue subject, CaseBlockList casesOrRules,
 			// Constructor is not "appl", or at least one of the patterns is not a concrete pattern
 			boolean hasChanged = false;
 			boolean hasMatched = false;
-			
+
+
+			if (cons.getConstructorType() == RascalValueFactory.Tree_Amb) {
```

Comment:
This fast trick could also be used in the deep match code (would save there some time avoiding `isSubtypeOf(Tree)`)

BTW: @PieterOlivier this design of integer/pointer-equal constructor nodes was taken from the ATerm library :-)